worker/src/socket.rs

```diff
@@ -0,0 +1,368 @@
+                    Err(e) => (Closing::None, Poll::Ready(Err(js_value_to_std_io_error(e)))),
+                },
+            }
+        }
+        let (new_closing, poll) = match self.close.take().unwrap_or_default() {
+            Closing::None => handle_future(cx, JsFuture::from(self.writer.close())),
+            Closing::Pending(fut) => handle_future(cx, fut),
+        };
+        self.close = Some(new_closing);
+        poll
+    }
+}
+
+pub enum SecureTransport {
+    /// Do not use TLS.
+    Off,
+    /// Use TLS.
+    On,
+    /// Do not use TLS initially, but allow the socket to be upgraded to
+    /// use TLS by calling [`Socket.start_tls`](Socket::start_tls).
+    StartTls,
+}
+
+pub struct SocketOptions {
+    /// Specifies whether or not to use TLS when creating the TCP socket.
+    pub secure_transport: SecureTransport,
+    /// Defines whether the writable side of the TCP socket will automatically
+    /// close on end-of-file (EOF). When set to false, the writable side of the
+    /// TCP socket will automatically close on EOF. When set to true, the
+    /// writable side of the TCP socket will remain open on EOF.
+    pub allow_half_open: bool,
+}
+
+impl Default for SocketOptions {
+    fn default() -> Self {
+        SocketOptions {
+            secure_transport: SecureTransport::Off,
+            allow_half_open: false,
+        }
+    }
+}
+
+/// The host and port that you wish to connect to.
+pub struct SocketAddress {
+    /// The hostname to connect to. Example: `cloudflare.com`.
+    pub hostname: String,
+    /// The port number to connect to. Example: `5432`.
+    pub port: u16,
+}
+
+#[derive(Default)]
+pub struct ConnectionBuilder {
+    hostname: Option<String>,
+    port: Option<u16>,
+    options: SocketOptions,
+}
+
+impl ConnectionBuilder {
+    /// Create a new `ConnectionBuilder` with default settings.
+    pub fn new() -> Self {
+        ConnectionBuilder {
+            hostname: None,
+            port: None,
+            options: SocketOptions::default(),
+        }
+    }
+
+    /// Set the hostname to connect to. Example: `cloudflare.com`.
+    pub fn host(mut self, host: &str) -> Self {
+        self.hostname = Some(host.to_string());
+        self
+    }
+
+    /// Set the port to connect to. Example: `5432`.
+    pub fn port(mut self, port: u16) -> Self {
+        self.port = Some(port);
+        self
+    }
+
+    /// Set whether the writable side of the TCP socket will automatically
+    /// close on end-of-file (EOF).
+    pub fn allow_half_open(mut self, allow_half_open: bool) -> Self {
+        self.options.allow_half_open = allow_half_open;
+        self
+    }
+
+    pub fn secure_transport(mut self, secure_transport: SecureTransport) -> Self {
+        self.options.secure_transport = secure_transport;
+        self
+    }
+
+    /// Open the connection to `hostname` on port `port`, returning a [`Socket`](Socket).
+    pub fn connect(self, hostname: &str, port: u16) -> Result<Socket> {
```

Comment:
Shouldn't this either use the `host` and `port` builder methods and not accept any args here or those methods not exist?